File: lelamp/tools.py
"""High-level LeLamp tools shared by local commands and external agents."""
from __future__ import annotations

from dataclasses import dataclass, field
from enum import StrEnum
from typing import Any, TYPE_CHECKING

if TYPE_CHECKING:
    from .app import LampApp


@dataclass(frozen=True)
class ToolOutcome:
    status: str
    message: str
    data: dict[str, Any] = field(default_factory=dict)

    def as_dict(self) -> dict[str, Any]:
        return {"ok": self.status not in {"failed", "cancelled"},
                "status": self.status, "message": self.message, "data": self.data}


class ToolSource(StrEnum):
    AGENT = "agent"
    LOCAL_VOICE = "local_voice"
    VISION = "vision"
    SCHEDULED = "scheduled"
    CONTROL_API = "control_api"


class ToolExecutor:
    """The only public path from intent sources into coordinated app actions."""

    MOTIONS = {
        "happy_wiggle", "excited", "sad", "shy", "shock",
        "nod", "headshake", "curious",
    }
    EXPRESSIONS = MOTIONS
    LIGHT_MODES = {"on", "off", "office", "warm"}
    SCHEDULED_TOOLS = {
        "play_motion", "set_light", "enter_work_light",
        "update_work_light", "exit_work_light", "sleep", "stop_tracking",
        "start_face_tracking",
        "turn_base", "reset_base_heading",
        "set_base_heading",
    }

    def __init__(self, app: "LampApp") -> None:
        self.app = app

    def _callback_info(
        self, args: dict[str, Any], source: ToolSource
    ) -> dict[str, Any]:
        callback_info: dict[str, Any] = {"source": source.value}
        on_complete = args.get("on_complete")
        agent_task = str(args.get("agent_task", "")).strip()
        if on_complete is not None and agent_task:
            raise ValueError("on_complete 和 agent_task 不能同时设置")
        if on_complete is not None:
            if not isinstance(on_complete, dict):
                raise ValueError("on_complete 必须是对象")
            tool_name = str(on_complete.get("tool", ""))
            tool_args = on_complete.get("arguments", {})
            if tool_name not in self.SCHEDULED_TOOLS:
                raise ValueError(f"不支持定时执行的工具: {tool_name}")
            if not isinstance(tool_args, dict):
                raise ValueError("on_complete.arguments 必须是对象")
            callback_info["action"] = {"tool": tool_name, "arguments": dict(tool_args)}
        elif agent_task:
            callback_info["agent_task"] = agent_task
        return callback_info

    async def execute(
        self,
        name: str,
        arguments: dict[str, Any] | None = None,
        *,
        source: ToolSource | str = ToolSource.CONTROL_API,
    ) -> ToolOutcome:
        try:
            source = ToolSource(source)
        except ValueError as exc:
            raise ValueError(f"未知工具调用来源: {source}") from exc
        args = arguments or {}
        if name == "create_timer":
            duration = args.get("duration_seconds", 0)
            message = str(args.get("message", ""))
            timer = await self.app.timers.create_timer(
                duration, message, self._callback_info(args, source)
            )
            return ToolOutcome("completed", "计时器已创建", timer.as_dict())

        if name == "create_alarm":
            alarm = await self.app.alarms.create_alarm(
                str(args.get("trigger_at", "")),
                str(args.get("message", "")),
                str(args.get("recurrence", "once")),
                self._callback_info(args, source),
                args.get("day_of_week"),
            )
            return ToolOutcome("completed", "闹钟已创建", alarm.as_dict())

        if name == "cancel_alarm":
            alarm = await self.app.alarms.cancel_alarm(str(args.get("alarm_id", "")))
            return ToolOutcome("completed", "闹钟已取消", alarm.as_dict())

        if name == "get_alarm":
            alarm = await self.app.alarms.get_alarm(str(args.get("alarm_id", "")))
            return ToolOutcome("completed", "闹钟状态读取成功", alarm.as_dict())

        if name == "list_alarms":
            alarms = await self.app.alarms.list_alarms(bool(args.get("include_finished", False)))
            return ToolOutcome(
                "completed", "闹钟列表读取成功",
                {"alarms": [alarm.as_dict() for alarm in alarms]},
            )

        if name == "pause_timer":
            timer = await self.app.timers.pause_timer(str(args.get("timer_id", "")))
            return ToolOutcome("completed", "计时器已暂停", timer.as_dict())

        if name == "resume_timer":
            timer = await self.app.timers.resume_timer(str(args.get("timer_id", "")))
            return ToolOutcome("completed", "计时器已继续", timer.as_dict())

        if name == "cancel_timer":
            timer = await self.app.timers.cancel_timer(str(args.get("timer_id", "")))
            return ToolOutcome("completed", "计时器已取消", timer.as_dict())

        if name == "add_timer_time":
            timer = await self.app.timers.add_time(
                str(args.get("timer_id", "")), args.get("seconds", 0)
            )
            return ToolOutcome("completed", "计时器已加时", timer.as_dict())

        if name == "get_timer_remaining":
            timer = await self.app.timers.get_remaining(str(args.get("timer_id", "")))
            return ToolOutcome("completed", "计时器状态读取成功", timer.as_dict())

        if name == "list_timers":
            timers = await self.app.timers.list_timers(
                bool(args.get("include_finished", False))
            )
            return ToolOutcome(
                "completed", "计时器列表读取成功",
                {"timers": [timer.as_dict() for timer in timers]},
            )

        if name == "play_motion":
            motion = str(args.get("name", ""))
            if motion not in self.MOTIONS:
                raise ValueError(f"不支持的动作: {motion}")
            if source is ToolSource.AGENT and self.app.should_defer_agent_motion(motion):
                queued, reason = self.app.queue_expression(motion)
                return ToolOutcome(
                    "completed",
                    "Agent 情绪动作已改为随本轮播报执行" if queued else reason,
                    {"motion": motion, "queued": queued, "timing": "with_reply"},
                )
            await self.app.play_motion(motion)
            return ToolOutcome("completed", "动作已完成", {"motion": motion})

        if name == "turn_base":
            direction = str(args.get("direction", ""))
            steps = args.get("steps", 1)
            if isinstance(steps, bool) or not isinstance(steps, int):
                raise ValueError("steps 必须是正整数")
            heading = await self.app.turn_base(direction, steps)
            return ToolOutcome(
                "completed", "底座转向已完成",
                {"direction": direction, "steps": steps,
                 "base_heading_degrees": heading},
            )

        if name == "reset_base_heading":
            heading = await self.app.reset_base_heading()
            return ToolOutcome(
                "completed", "底座已回到初始正面",
                {"base_heading_degrees": heading},
            )

        if name == "set_base_heading":
            position = str(args.get("position", ""))
            heading = await self.app.set_base_heading(position)
            return ToolOutcome(
                "completed", "底座绝对朝向已设置",
                {"position": position, "base_heading_degrees": heading},
            )

        if name == "queue_expression":
            if source is not ToolSource.AGENT:
                raise ValueError("queue_expression 只允许 Agent 自主表达调用")
            expression = str(args.get("name", ""))
            if expression not in self.EXPRESSIONS:
                raise ValueError(f"不支持的情绪动作: {expression}")
            queued, reason = self.app.queue_expression(expression)
            return ToolOutcome(
                "completed",
                "情绪动作已加入本轮播报" if queued else reason,
                {"expression": expression, "queued": queued},
            )

        if name == "set_light":
            mode = str(args.get("mode", ""))
            if mode not in self.LIGHT_MODES:
                raise ValueError(f"不支持的灯光模式: {mode}")
            brightness = args.get("brightness")
            if brightness is not None:
                brightness = float(brightness)
                if not 0 <= brightness <= 100:
                    raise ValueError("brightness 必须在 0～100")
            if mode == "off":
                self.app.set_light((0, 0, 0), 0)
            elif mode in {"on", "office"}:
                self.app.lighting.office_mode(brightness)
            else:
                self.app.lighting.warm_mode(brightness)
            return ToolOutcome("completed", "灯光已设置", {"mode": mode})

        if name == "enter_work_light":
            pose = str(args.get("pose", "high"))
            tone = str(args.get("tone", "white"))
            await self.app.enter_work_light(pose, tone)
            return ToolOutcome(
                "completed", "已进入办公照明模式",
                {"pose": pose, "tone": tone, "brightness": self.app.work_brightness},
            )

        if name == "update_work_light":
            pose = args.get("pose")
            tone = args.get("tone")
            step = args.get("brightness_step")
            if step is not None:
                step = int(step)
            await self.app.update_work_light(pose, tone, step)
            return ToolOutcome(
                "completed", "办公照明已调整",
                {
                    "pose": self.app.work_pose,
                    "tone": self.app.work_tone,
                    "brightness": self.app.work_brightness,
                },
            )

        if name == "exit_work_light":
            exited = await self.app.exit_work_light()
            return ToolOutcome(
                "completed",
                "已退出办公照明模式" if exited else "当前未处于办公照明模式",
                {"exited": exited},
            )

        if name == "sleep":
            reason = str(args.get("reason", "user_request"))
            self.app.request_shutdown(reason)
            return ToolOutcome(
                "shutdown_requested",
                "已登记休眠请求；请先完成简短告别，设备随后休眠",
                {"reason": reason},
            )

        if name == "get_robot_state":
            return ToolOutcome("completed", "状态读取成功", self.app.get_robot_state())

        if name == "get_vision_state":
            return ToolOutcome(
                "completed", "视觉状态读取成功", self.app.get_vision_state()
            )

        if name == "get_system_health":
            from .diagnostics import run_live
            return ToolOutcome("completed", "只读自检完成", await run_live(self.app))

        if name == "stop_tracking":
            was_tracking = self.app.current_mode == "tracking"
            await self.app.stop_tracking()
            return ToolOutcome(
                "completed",
                "跟踪已停止" if was_tracking else "当前没有运行跟踪",
                {"was_tracking": was_tracking},
            )

        if name == "start_face_tracking":
            await self.app.start_face_tracking()
            return ToolOutcome(
                "completed", "已开始人脸跟踪", {"mode": "tracking"}
            )

        raise ValueError(f"未知工具: {name}")

File: pi-agent/src/tools.ts
import type { AgentTool } from "@earendil-works/pi-agent-core";
import { Type, type TSchema } from "@earendil-works/pi-ai";
import { SearchRouter } from "./search.js";

export type ToolCallMetric = {
  name: string;
  arguments: Record<string, unknown>;
  duration_ms: number;
  ok: boolean;
};

export type RequestState = { toolCalls: ToolCallMetric[]; searchCount: number };

const expression = Type.Union([
  Type.Literal("happy_wiggle"), Type.Literal("excited"), Type.Literal("sad"),
  Type.Literal("shy"), Type.Literal("shock"), Type.Literal("nod"),
  Type.Literal("headshake"), Type.Literal("curious"),
]);
const scheduledAction = Type.Object({
  tool: Type.Union([
    Type.Literal("play_motion"), Type.Literal("set_light"),
    Type.Literal("enter_work_light"), Type.Literal("update_work_light"),
    Type.Literal("exit_work_light"), Type.Literal("sleep"),
    Type.Literal("stop_tracking"), Type.Literal("turn_base"),
    Type.Literal("reset_base_heading"), Type.Literal("set_base_heading"),
  ]),
  arguments: Type.Optional(Type.Record(Type.String(), Type.Unknown())),
}, { additionalProperties: false });

function result(value: unknown) {
  return { content: [{ type: "text" as const, text: JSON.stringify(value) }], details: value };
}

function lampTool(
  name: string, label: string, description: string, parameters: TSchema,
  controlName: string, request: RequestState,
): AgentTool {
  return {
    name, label, description, parameters,
    executionMode: "sequential",
    async execute(_id, params, signal) {
      const started = performance.now();
      let ok = false;
      try {
        const base = (process.env.LELAMP_CONTROL_URL ?? "http://127.0.0.1:18790").replace(/\/$/, "");
        const token = process.env.LELAMP_CONTROL_TOKEN ?? "";
        const response = await fetch(`${base}/v1/tools/${controlName}`, {
          method: "POST", signal,
          headers: { Authorization: `Bearer ${token}`, "Content-Type": "application/json" },
          body: JSON.stringify(params),
        });
        const body = await response.json() as Record<string, unknown>;
        if (!response.ok || body.ok === false) {
          throw new Error(String(body.message ?? `LeLamp HTTP ${response.status}`));
        }
        ok = true;
        return result(body);
      } finally {
        request.toolCalls.push({
          name, arguments: params as Record<string, unknown>,
          duration_ms: performance.now() - started, ok,
        });
      }
    },
  };
}

export const PI_TOOL_CONTRACTS = [
  "lelamp_web_search", "lelamp_play_motion", "lelamp_queue_expression",
  "lelamp_turn_base", "lelamp_reset_base_heading", "lelamp_set_base_heading",
  "lelamp_set_light", "lelamp_enter_work_light", "lelamp_update_work_light",
  "lelamp_exit_work_light", "lelamp_sleep", "lelamp_create_timer",
  "lelamp_pause_timer", "lelamp_resume_timer", "lelamp_cancel_timer",
  "lelamp_add_timer_time", "lelamp_get_timer_remaining", "lelamp_list_timers",
  "lelamp_create_alarm", "lelamp_cancel_alarm", "lelamp_get_alarm",
  "lelamp_list_alarms", "lelamp_get_robot_state",
] as const;

export function createTools(request: RequestState, search: SearchRouter): AgentTool[] {
  const lamp = (name: string, label: string, description: string, schema: TSchema, target?: string) =>
    lampTool(name, label, description, schema, target ?? name.replace(/^lelamp_/, ""), request);
  return [
    {
      name: "lelamp_web_search", label: "Search the web",
      description: "Search current web information. Use for weather, news, traffic, prices, schedules and other changing facts. Search at most four times per user turn.",
      parameters: Type.Object({ query: Type.String({ minLength: 1, maxLength: 500 }) }, { additionalProperties: false }),
      executionMode: "sequential",
      async execute(_id, params, signal) {
        const started = performance.now();
        let ok = false;
        try {
          if (request.searchCount >= 4) throw new Error("本轮联网搜索已达到 4 次上限，请根据已有结果回答");
          request.searchCount += 1;
          const body = await search.search(String((params as { query: string }).query), signal);
          ok = true;
          return result(body);
        } finally {
          request.toolCalls.push({ name: "lelamp_web_search", arguments: params as Record<string, unknown>, duration_ms: performance.now() - started, ok });
        }
      },
    },
    lamp("lelamp_play_motion", "Play motion", "用户明确命令台灯点头、摇头或做其他动作时必须调用此工具，而且只调用一次。包括带少量 ASR 错字但动作意图明确的请求。显式动作绝不能改用 lelamp_queue_expression。", Type.Object({ name: expression }, { additionalProperties: false })),
    lamp("lelamp_queue_expression", "Queue expression", "仅用于台灯对聊天内容自主表达情绪，绝不能执行用户明确要求的点头、摇头等动作，也不能在其他实体工具成功后追加。夸奖用 shy，认同用 nod，否定用 headshake，开心用 happy_wiggle，兴奋用 excited，安慰用 sad，惊讶用 shock，好奇用 curious。", Type.Object({ name: expression }, { additionalProperties: false })),
    lamp("lelamp_turn_base", "Turn base", "Turn left or right in 30-degree relative steps.", Type.Object({ direction: Type.Union([Type.Literal("left"), Type.Literal("right")]), steps: Type.Optional(Type.Integer({ minimum: 1, maximum: 2 })) }, { additionalProperties: false })),
    lamp("lelamp_reset_base_heading", "Reset heading", "Return the base to the original front heading.", Type.Object({}, { additionalProperties: false })),
    lamp("lelamp_set_base_heading", "Set absolute heading", "Set the absolute left, front, or right heading.", Type.Object({ position: Type.Union([Type.Literal("left"), Type.Literal("front"), Type.Literal("right")]) }, { additionalProperties: false })),
    lamp("lelamp_set_light", "Set light", "Set light off, office neutral-white, warm yellow, or on alias.", Type.Object({ mode: Type.Union([Type.Literal("on"), Type.Literal("off"), Type.Literal("office"), Type.Literal("warm")]), brightness: Type.Optional(Type.Number({ minimum: 0, maximum: 100 })) }, { additionalProperties: false })),
    lamp("lelamp_enter_work_light", "Enter work light", "Start persistent desk illumination; default high pose, white tone, 75 percent.", Type.Object({ pose: Type.Optional(Type.Union([Type.Literal("high"), Type.Literal("low")])), tone: Type.Optional(Type.Union([Type.Literal("white"), Type.Literal("warm")])) }, { additionalProperties: false })),
    lamp("lelamp_update_work_light", "Adjust work light", "Adjust active work-light pose, tone, or brightness by exactly 25 points.", Type.Object({ pose: Type.Optional(Type.Union([Type.Literal("high"), Type.Literal("low")])), tone: Type.Optional(Type.Union([Type.Literal("white"), Type.Literal("warm")])), brightness_step: Type.Optional(Type.Union([Type.Literal(-25), Type.Literal(25)])) }, { additionalProperties: false })),
    lamp("lelamp_exit_work_light", "Exit work light", "End work-light mode and return to normal standby.", Type.Object({}, { additionalProperties: false })),
    lamp("lelamp_sleep", "Sleep", "Request mechanical sleep for a direct farewell or sleep request, never quotation, question, or negation.", Type.Object({ reason: Type.Union([Type.Literal("farewell"), Type.Literal("user_request")]) }, { additionalProperties: false })),
    lamp("lelamp_create_timer", "Create timer", "创建计时器。普通提醒只填写 message；用户明确要求到期执行固定机器人动作才填写 on_complete；到期时才推理则只填 agent_task。条件请求必须先读取查询或状态工具的实际结果，只有条件明确为 true 才能调用；false 或 unknown 时严禁调用。", Type.Object({ duration_seconds: Type.Number({ exclusiveMinimum: 0 }), message: Type.String(), on_complete: Type.Optional(scheduledAction), agent_task: Type.Optional(Type.String({ minLength: 1, maxLength: 2000 })) }, { additionalProperties: false })),
    lamp("lelamp_pause_timer", "Pause timer", "Pause a running timer.", Type.Object({ timer_id: Type.String() }, { additionalProperties: false })),
    lamp("lelamp_resume_timer", "Resume timer", "Resume a paused timer.", Type.Object({ timer_id: Type.String() }, { additionalProperties: false })),
    lamp("lelamp_cancel_timer", "Cancel timer", "Cancel a timer without firing completion.", Type.Object({ timer_id: Type.String() }, { additionalProperties: false })),
    lamp("lelamp_add_timer_time", "Add timer time", "Add positive seconds to a running or paused timer.", Type.Object({ timer_id: Type.String(), seconds: Type.Number({ exclusiveMinimum: 0 }) }, { additionalProperties: false })),
    lamp("lelamp_get_timer_remaining", "Get timer", "Get one timer status and remaining seconds.", Type.Object({ timer_id: Type.String() }, { additionalProperties: false })),
    lamp("lelamp_list_timers", "List timers", "List timers in creation order.", Type.Object({ include_finished: Type.Optional(Type.Boolean()) }, { additionalProperties: false })),
    lamp("lelamp_create_alarm", "Create alarm", "Create an absolute-time alarm. trigger_at uses device timezone offset; recurrence is once, daily, weekdays, or weekly. Plain reminders use message only; on_complete and agent_task are mutually exclusive.", Type.Object({ trigger_at: Type.String({ minLength: 20 }), recurrence: Type.Union([Type.Literal("once"), Type.Literal("daily"), Type.Literal("weekdays"), Type.Literal("weekly")]), day_of_week: Type.Optional(Type.Integer({ minimum: 1, maximum: 7 })), message: Type.String(), on_complete: Type.Optional(scheduledAction), agent_task: Type.Optional(Type.String({ minLength: 1, maxLength: 2000 })) }, { additionalProperties: false })),
    lamp("lelamp_cancel_alarm", "Cancel alarm", "Cancel a scheduled alarm.", Type.Object({ alarm_id: Type.String() }, { additionalProperties: false })),
    lamp("lelamp_get_alarm", "Get alarm", "Read an alarm and its next trigger.", Type.Object({ alarm_id: Type.String() }, { additionalProperties: false })),
    lamp("lelamp_list_alarms", "List alarms", "List scheduled alarms in creation order.", Type.Object({ include_finished: Type.Optional(Type.Boolean()) }, { additionalProperties: false })),
    lamp("lelamp_get_robot_state", "Get robot state", "Read coordinated high-level state. base_heading_position is authoritative.", Type.Object({}, { additionalProperties: false })),
  ];
}
